### PR TITLE
add Azure support

### PR DESCRIPTION
**Pull request by kdomanski**
**Release note**:
```release-note cloud provider
Added Azure support
```

### CODE REVIEW

File: api/pkg/provider/cloud/azure/provider.go [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+	}
+
+	parameters := resources.Group{
+		Name:     to.StringPtr(cloud.Azure.ResourceGroup),
+		Location: to.StringPtr(location),
+		Tags: map[string]*string{
+			clusterTagKey: to.StringPtr(clusterName),
+		},
+	}
+	if _, err = groupsClient.CreateOrUpdate(context.TODO(), cloud.Azure.ResourceGroup, parameters); err != nil {
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go:172]: Using `CreateOrUpdate` might be a bit confusing if the function is called `createResourceGroup`.

File: api/pkg/provider/cloud/azure/provider.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	groupsClient, err := getGroupsClient(cloud)
+	if err != nil {
+		return err
+	}
+
+	parameters := resources.Group{
+		Name:     to.StringPtr(cloud.Azure.ResourceGroup),
+		Location: to.StringPtr(location),
+		Tags: map[string]*string{
+			clusterTagKey: to.StringPtr(clusterName),
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go:169]: Does this override existing tags? Or is it a merge operation?

File: api/pkg/provider/cloud/azure/provider.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+func createVNet(cloud *kubermaticv1.CloudSpec, clusterName string) error {
+	networksClient, err := getNetworksClient(cloud)
+	if err != nil {
+		return err
+	}
+
+	parameters := network.VirtualNetwork{
+		Name: to.StringPtr(cloud.Azure.VNetName),
+		Tags: map[string]*string{
+			clusterTagKey: to.StringPtr(clusterName),
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go]: Does this override existing tags? Or is it a merge operation?

Overriding is probably very destructive & might break scenarios where multiple clusters share the same network?

File: api/pkg/provider/cloud/azure/provider.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+		},
+	}
+
+	future, err := networksClient.CreateOrUpdate(context.TODO(), cloud.Azure.ResourceGroup, cloud.Azure.VNetName, parameters)
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go]: Using `CreateOrUpdate` might be a bit confusing if the function is called `createVNet`.

File: api/pkg/provider/cloud/azure/provider.go [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+	if err := deleteRouteTable(cloud); err != nil {
+		return err
+	}
+
+	if err := deleteSubnet(cloud); err != nil {
+		return err
+	}
+
+	if err := deleteVNet(cloud); err != nil {
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go]: I guess, if we always delete the resources, the option that the user can specify some things, is not given with this PR.
That is perfectly fine for now - it's something we can do in a follow-up PR.

File: api/pkg/provider/cloud/azure/provider.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/provider/cloud/azure/provider.go:169]: Not relevant due to https://github.com/kubermatic/kubermatic/pull/1200/files#r193963266

File: api/pkg/provider/cloud/azure/provider.go (hunk 3 above)
Comment (mrIncompetent) [reply api/pkg/provider/cloud/azure/provider.go]: Not relevant due to https://github.com/kubermatic/kubermatic/pull/1200/files#r193963266

File: api/pkg/provider/cloud/azure/provider.go [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+	subnetsClient, err := getSubnetsClient(cloud)
+	if err != nil {
+		return err
+	}
+
+	parameters := network.Subnet{
+		Name: to.StringPtr(cloud.Azure.SubnetName),
+	}
+
+	future, err := subnetsClient.CreateOrUpdate(context.TODO(), cloud.Azure.ResourceGroup, cloud.Azure.VNetName, cloud.Azure.SubnetName, parameters)
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go]: Using `CreateOrUpdate` might be a bit confusing if the function is called `createSubnet`.

File: api/pkg/provider/cloud/azure/provider.go [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+		RouteTablePropertiesFormat: &network.RouteTablePropertiesFormat{
+			Subnets: &[]network.Subnet{
+				network.Subnet{
+					Name: to.StringPtr(cloud.Azure.SubnetName),
+				},
+			},
+		},
+	}
+
+	future, err := routeTablesClient.CreateOrUpdate(context.TODO(), cloud.Azure.ResourceGroup, cloud.Azure.RouteTableName, parameters)
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go:366]: Using `CreateOrUpdate` might be a bit confusing if the function is called `createRouteTable`.

File: api/pkg/provider/cloud/azure/provider.go [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+	if err := createSubnet(cloud); err != nil {
+		return nil, err
+	}
+
+	if err := createRouteTable(cloud); err != nil {
+		return nil, err
+	}
+
+	// TODO create security group
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go]: ?

File: api/pkg/provider/cloud/azure/provider.go [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+
+	if err = future.WaitForCompletion(context.TODO(), routeTablesClient.Client); err != nil {
+		return fmt.Errorf("failed to delete route table %q: %v", cloud.Azure.RouteTableName, err)
+	}
+
+	return nil
+}
+
+func (a *azure) CleanUpCloudProvider(cloud *kubermaticv1.CloudSpec) error {
+	// TODO delete security group
```
Comment (mrIncompetent) [inline api/pkg/provider/cloud/azure/provider.go]: ?

**Review by mrIncompetent — CHANGES_REQUESTED** [review]
Just some nits

File: api/pkg/provider/cloud/azure/provider.go (hunk 1 above)
Comment (kdomanski) [reply api/pkg/provider/cloud/azure/provider.go:172]: done.

File: api/pkg/provider/cloud/azure/provider.go (hunk 4 above)
Comment (kdomanski) [reply api/pkg/provider/cloud/azure/provider.go]: done

File: api/pkg/provider/cloud/azure/provider.go (hunk 6 above)
Comment (kdomanski) [reply api/pkg/provider/cloud/azure/provider.go]: done.

File: api/pkg/provider/cloud/azure/provider.go (hunk 7 above)
Comment (kdomanski) [reply api/pkg/provider/cloud/azure/provider.go:366]: done.

File: api/pkg/provider/cloud/azure/provider.go (hunk 8 above)
Comment (kdomanski) [reply api/pkg/provider/cloud/azure/provider.go]: done.

File: api/pkg/provider/cloud/azure/provider.go (hunk 9 above)
Comment (kdomanski) [reply api/pkg/provider/cloud/azure/provider.go]: done

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**